File: drivers/php/src/RocheDB.php
<?php
declare(strict_types=1);

namespace RocheDB;

use FFI;
use FFI\CData;
use RuntimeException;

final class RocheId
{
    public function __construct(
        public readonly int|string $parent,
        public readonly int $epoch,
        public readonly int $seq,
        public readonly float $tWrite,
    ) {
    }
}

final class RocheHit
{
    public function __construct(
        public readonly RocheId $id,
        public readonly float $score,
        public readonly string $payload,
    ) {
    }
}

final class RetrieveResult
{
    /** @param list<RocheHit> $hits */
    public function __construct(
        public readonly array $hits,
        public readonly array $stats,
    ) {
    }
}

final class RocheDB
{
    private const ABI_VERSION = 1;

    private static ?FFI $ffi = null;
    private ?CData $handle;

    private function __construct(CData $handle)
    {
        $this->handle = $handle;
    }

    public function __destruct()
    {
        $this->close();
    }

    public static function open(int $nodes = 8, string $lib = __DIR__ . '/../../../lib/librochedb.so'): self
    {
        $ffi = self::ffi($lib);
        $handle = $ffi->roche_open($nodes);
        if ($handle === null) {
            throw self::lastError();
        }
        return new self($handle);
    }

    public static function openDir(int $nodes, string $dir, string $lib = __DIR__ . '/../../../lib/librochedb.so'): self
    {
        $ffi = self::ffi($lib);
        $handle = $ffi->roche_open_dir($nodes, $dir);
        if ($handle === null) {
            throw self::lastError();
        }
        return new self($handle);
    }

    public static function connectAuth(
        string $peers,
        string $username = '',
        string $password = '',
        string $authToken = '',
        string $secretKey = '',
        string $galaxy = '',
        string $lib = __DIR__ . '/../../../lib/librochedb.so',
    ): self {
        $ffi = self::ffi($lib);
        $handle = $ffi->roche_connect_auth($peers, $username, $password, $authToken, $secretKey, $galaxy);
        if ($handle === null) {
            throw self::lastError();
        }
        return new self($handle);
    }

    public static function connect(string $peers, string $lib = __DIR__ . '/../../../lib/librochedb.so'): self
    {
        return self::connectAuth($peers, lib: $lib);
    }

    public function close(): void
    {
        if ($this->handle !== null) {
            self::ffi()->roche_close($this->handle);
            $this->handle = null;
        }
    }

    public function configureRing(string $ring, float $period): void
    {
        $this->check(self::ffi()->roche_ring_configure($this->requireHandle(), $ring, $period));
    }

    public function setGalaxyDescription(string $description): void
    {
        $this->check(self::ffi()->roche_set_galaxy_description($this->requireHandle(), $description));
    }

    public function setRingDescription(string $ring, string $description): void
    {
        $this->check(self::ffi()->roche_set_ring_description($this->requireHandle(), $ring, $description));
    }

    public function put(string $ring, string $payload): RocheId
    {
        $ffi = self::ffi();
        $id = $ffi->new('roche_id');
        $this->check($ffi->roche_put($this->requireHandle(), $ring, $payload, strlen($payload), FFI::addr($id)));
        return self::idFromC($id);
    }

    /** @param list<float|int> $vector */
    public function putVec(string $ring, string $payload, array $vector): RocheId
    {
        $ffi = self::ffi();
        $id = $ffi->new('roche_id');
        $vec = self::floatArray($vector);
        $vecPtr = count($vector) === 0 ? null : FFI::addr($vec[0]);
        $this->check($ffi->roche_put_vec(
            $this->requireHandle(),
            $ring,
            $payload,
            strlen($payload),
            $vecPtr,
            count($vector),
            FFI::addr($id),
        ));
        return self::idFromC($id);
    }

    public function get(RocheId $id): ?string
    {
        $ffi = self::ffi();
        $len = $ffi->new('size_t[1]');
        $ptr = $ffi->roche_get($this->requireHandle(), self::idToC($id), FFI::addr($len[0]));
        if ($ptr === null) {
            $message = self::lastError()->getMessage();
            if (str_contains($message, 'not found')) {
                return null;
            }
            throw new RuntimeException($message);
        }
        try {
            return FFI::string($ptr, (int)$len[0]);
        } finally {
            $ffi->roche_free($ptr);
        }
    }

    /** @param list<RocheId> $ids @return list<string> */
    public function batchGet(array $ids): array
    {
        $ffi = self::ffi();
        $arr = $ffi->new('roche_id[' . max(1, count($ids)) . ']');
        foreach ($ids as $i => $id) {
            $arr[$i] = self::idToC($id);
        }
        $res = $ffi->roche_batch_get($this->requireHandle(), count($ids) === 0 ? null : FFI::addr($arr[0]), count($ids));
        if ($res === null) {
            throw self::lastError();
        }
        try {
            $out = [];
            for ($i = 0; $i < (int)$res->len; $i++) {
                $value = $res->values[$i];
                $out[] = $value->data === null ? '' : FFI::string($value->data, (int)$value->len);
            }
            return $out;
        } finally {
            $ffi->roche_batch_get_free($res);
        }
    }

    public function query(RocheId $id, string $selection): string
    {
        $ffi = self::ffi();
        $len = $ffi->new('size_t[1]');
        $ptr = $ffi->roche_query($this->requireHandle(), self::idToC($id), $selection, FFI::addr($len[0]));
        if ($ptr === null) {
            throw self::lastError();
        }
        try {
            return FFI::string($ptr, (int)$len[0]);
        } finally {
            $ffi->roche_free($ptr);
        }
    }

    /** @param list<float|int> $vector */
    public function retrieve(array $vector, string $ring = '', int $budget = 8, int $topRings = 0, int $focus = 0): RetrieveResult
    {
        $ffi = self::ffi();
        $vec = self::floatArray($vector);
        $res = $ffi->roche_retrieve(
            $this->requireHandle(),
            count($vector) === 0 ? null : FFI::addr($vec[0]),
            count($vector),
            $ring,
            $budget,
            $topRings,
            $focus,
        );
        if ($res === null) {
            throw self::lastError();
        }
        try {
            $hits = [];
            for ($i = 0; $i < (int)$res->len; $i++) {
                $hit = $res->hits[$i];
                $hits[] = new RocheHit(
                    self::idFromC($hit->id),
                    (float)$hit->score,
                    $hit->payload === null ? '' : FFI::string($hit->payload, (int)$hit->payload_len),
                );
            }
            return new RetrieveResult($hits, [
                'totalVectors' => (int)$res->total_vectors,
                'scanned' => (int)$res->scanned,
                'skippedVectors' => (int)$res->skipped_vectors,
                'returned' => (int)$res->returned,
                'ringsTouched' => (int)$res->rings_touched,
                'payloadBytes' => (int)$res->payload_bytes,
                'estimatedTokens' => (int)$res->estimated_tokens,
                'fanoutNodes' => (int)$res->fanout_nodes,
                'candidateReduction' => (float)$res->candidate_reduction,
            ]);
        } finally {
            $ffi->roche_retrieve_free($res);
        }
    }

    /** @param list<float|int> $queryVector */
    public function atlas(array $queryVector = [], int $maxCentroidDims = 8): string
    {
        $ffi = self::ffi();
        $vec = self::floatArray($queryVector);
        $len = $ffi->new('size_t[1]');
        $ptr = $ffi->roche_atlas(
            $this->requireHandle(),
            count($queryVector) === 0 ? null : FFI::addr($vec[0]),
            count($queryVector),
            $maxCentroidDims,
            FFI::addr($len[0]),
        );
        if ($ptr === null) {
            throw self::lastError();
        }
        try {
            return FFI::string($ptr, (int)$len[0]);
        } finally {
            $ffi->roche_free($ptr);
        }
    }

    public function locate(RocheId $id, float $at = -1.0): int
    {
        $node = self::ffi()->roche_locate($this->requireHandle(), self::idToC($id), $at);
        if ($node < 0) {
            throw self::lastError();
        }
        return (int)$node;
    }

    public function nextVisit(RocheId $id, int $node): float
    {
        $time = self::ffi()->roche_next_visit($this->requireHandle(), self::idToC($id), $node);
        if ($time < 0) {
            throw self::lastError();
        }
        return (float)$time;
    }

    public function nextJoin(RocheId $a, RocheId $b): ?float
    {
        $time = self::ffi()->roche_next_join($this->requireHandle(), self::idToC($a), self::idToC($b));
        return $time < 0 ? null : (float)$time;
    }

    private function check(int $code): void
    {
        if ($code !== 0) {
            throw self::lastError();
        }
    }

    private function requireHandle(): CData
    {
        if ($this->handle === null) {
            throw new RuntimeException('RocheDB handle is closed');
        }
        return $this->handle;
    }

    /** @param list<float|int> $values */
    private static function floatArray(array $values): CData
    {
        $ffi = self::ffi();
        $arr = $ffi->new('float[' . max(1, count($values)) . ']');
        foreach ($values as $i => $value) {
            $arr[$i] = (float)$value;
        }
        return $arr;
    }

    private static function idToC(RocheId $id): CData
    {
        $c = self::ffi()->new('roche_id');
        $c->parent = $id->parent;
        $c->epoch = $id->epoch;
        $c->seq = $id->seq;
        $c->t_write = $id->tWrite;
        return $c;
    }

    private static function idFromC(CData $id): RocheId
    {
        return new RocheId((string)$id->parent, (int)$id->epoch, (int)$id->seq, (float)$id->t_write);
    }

    private static function lastError(): RuntimeException
    {
        $ptr = self::ffi()->roche_last_error();
        $message = $ptr === null ? 'RocheDB C ABI error' : FFI::string($ptr);
        return new RuntimeException($message === '' ? 'RocheDB C ABI error' : $message);
    }

    private static function ffi(string $lib = __DIR__ . '/../../../lib/librochedb.so'): FFI
    {
        if (!class_exists(FFI::class)) {
            throw new RuntimeException('PHP FFI extension is not enabled');
        }
        if (self::$ffi === null) {
            self::$ffi = FFI::cdef(self::CDEF, $lib);
            self::$ffi->roche_init();
            if ((int)self::$ffi->roche_abi_version() !== self::ABI_VERSION) {
                throw new RuntimeException('RocheDB ABI version mismatch');
            }
        }
        return self::$ffi;
    }

    private const CDEF = <<<'CDEF'
typedef unsigned long size_t;
typedef unsigned long uint64_t;
typedef unsigned int uint32_t;

typedef struct roche_id {
  uint64_t parent;
  uint32_t epoch;
  uint32_t seq;
  double   t_write;
} roche_id;

typedef struct roche_hit {
  roche_id id;
  double   score;
  void    *payload;
  size_t   payload_len;
} roche_hit;

typedef struct roche_retrieve_result {
  size_t     len;
  roche_hit *hits;
  int        total_vectors;
  int        scanned;
  int        skipped_vectors;
  int        returned;
  int        rings_touched;
  int        payload_bytes;
  int        estimated_tokens;
  int        fanout_nodes;
  double     candidate_reduction;
} roche_retrieve_result;

typedef struct roche_value {
  void  *data;
  size_t len;
} roche_value;

typedef struct roche_batch_result {
  size_t       len;
  roche_value *values;
} roche_batch_result;

int         roche_abi_version(void);
const char *roche_last_error(void);
void        roche_init(void);
void       *roche_open(int nodes);
void       *roche_open_dir(int nodes, const char *dir);
void       *roche_connect_auth(const char *peers, const char *username, const char *password, const char *auth_token, const char *secret_key, const char *galaxy);
void        roche_close(void *db);
int         roche_ring_configure(void *db, const char *ring, double period);
int         roche_set_galaxy_description(void *db, const char *description);
int         roche_set_ring_description(void *db, const char *ring, const char *description);
int         roche_put(void *db, const char *ring, const void *data, size_t len, roche_id *out_id);
int         roche_put_vec(void *db, const char *ring, const void *data, size_t len, const float *vec, size_t vec_len, roche_id *out_id);
void       *roche_get(void *db, roche_id id, size_t *out_len);
void        roche_free(void *p);
roche_batch_result *roche_batch_get(void *db, const roche_id *ids, size_t ids_len);
void        roche_batch_get_free(roche_batch_result *r);
void       *roche_query(void *db, roche_id id, const char *selection, size_t *out_len);
roche_retrieve_result *roche_retrieve(void *db, const float *vec, size_t vec_len, const char *ring, int budget, int top_rings, int focus);
void        roche_retrieve_free(roche_retrieve_result *r);
void       *roche_atlas(void *db, const float *query_vec, size_t query_vec_len, int max_centroid_dims, size_t *out_len);
int         roche_locate(void *db, roche_id id, double at);
double      roche_next_visit(void *db, roche_id id, int node);
double      roche_next_join(void *db, roche_id a, roche_id b);
CDEF;
}

File: drivers/swift/Sources/RocheDB/RocheDB.swift
import CRocheDB
import Foundation

public enum RocheDBError: Error, CustomStringConvertible, Equatable {
    case abiVersionMismatch(expected: Int32, actual: Int32)
    case cAbi(String)
    case closed
    case invalidUtf8

    public var description: String {
        switch self {
        case let .abiVersionMismatch(expected, actual):
            return "RocheDB ABI version mismatch: expected \(expected), got \(actual)"
        case let .cAbi(message):
            return message
        case .closed:
            return "RocheDB handle is closed"
        case .invalidUtf8:
            return "RocheDB returned invalid UTF-8"
        }
    }
}

public struct RocheId: Equatable, Sendable {
    public let parent: UInt64
    public let epoch: UInt32
    public let seq: UInt32
    public let tWrite: Double

    public init(parent: UInt64, epoch: UInt32, seq: UInt32, tWrite: Double) {
        self.parent = parent
        self.epoch = epoch
        self.seq = seq
        self.tWrite = tWrite
    }
}

public struct RocheHit: Equatable, Sendable {
    public let id: RocheId
    public let score: Double
    public let payload: Data
}

public struct RetrieveStats: Equatable, Sendable {
    public let totalVectors: Int
    public let scanned: Int
    public let skippedVectors: Int
    public let returned: Int
    public let ringsTouched: Int
    public let payloadBytes: Int
    public let estimatedTokens: Int
    public let fanoutNodes: Int
    public let candidateReduction: Double
}

public struct RetrieveResult: Equatable, Sendable {
    public let hits: [RocheHit]
    public let stats: RetrieveStats
}

public final class RocheDB {
    public static let abiVersion: Int32 = 2
    private static let runtimeInit: Void = {
        roche_init()
    }()

    private var handle: UnsafeMutableRawPointer?

    private init(handle: UnsafeMutableRawPointer) {
        self.handle = handle
    }

    deinit {
        close()
    }

    public static func open(nodes: Int32 = 8) throws -> RocheDB {
        try ensureRuntime()
        guard let handle = roche_open(nodes) else {
            throw lastError()
        }
        return RocheDB(handle: handle)
    }

    public static func openDir(nodes: Int32 = 8, dir: String) throws -> RocheDB {
        try ensureRuntime()
        return try dir.withCString { cDir in
            guard let handle = roche_open_dir(nodes, cDir) else {
                throw lastError()
            }
            return RocheDB(handle: handle)
        }
    }

    public static func connectAuth(
        peers: String,
        username: String = "",
        password: String = "",
        authToken: String = "",
        secretKey: String = "",
        galaxy: String = ""
    ) throws -> RocheDB {
        try ensureRuntime()
        return try peers.withCString { cPeers in
            try username.withCString { cUsername in
                try password.withCString { cPassword in
                    try authToken.withCString { cAuthToken in
                        try secretKey.withCString { cSecretKey in
                            try galaxy.withCString { cGalaxy in
                                guard let handle = roche_connect_auth(
                                    cPeers,
                                    cUsername,
                                    cPassword,
                                    cAuthToken,
                                    cSecretKey,
                                    cGalaxy
                                ) else {
                                    throw lastError()
                                }
                                return RocheDB(handle: handle)
                            }
                        }
                    }
                }
            }
        }
    }

    public static func connect(peers: String) throws -> RocheDB {
        try connectAuth(peers: peers)
    }

    public func close() {
        if let handle {
            roche_close(handle)
            self.handle = nil
        }
    }

    public func configureRing(_ ring: String, period: Double) throws {
        try withHandle { handle in
            try ring.withCString { cRing in
                try check(roche_ring_configure(handle, cRing, period))
            }
        }
    }

    public func setGalaxyDescription(_ description: String) throws {
        try withHandle { handle in
            try description.withCString { cDescription in
                try check(roche_set_galaxy_description(handle, cDescription))
            }
        }
    }

    public func setRingDescription(_ ring: String, description: String) throws {
        try withHandle { handle in
            try ring.withCString { cRing in
                try description.withCString { cDescription in
                    try check(roche_set_ring_description(handle, cRing, cDescription))
                }
            }
        }
    }

    public func put(_ ring: String, payload: Data) throws -> RocheId {
        try withHandle { handle in
            try ring.withCString { cRing in
                var id = roche_id()
                let rc = payload.withUnsafeBytes { bytes in
                    roche_put(handle, cRing, bytes.baseAddress, payload.count, &id)
                }
                try check(rc)
                return RocheId(c: id)
            }
        }
    }

    public func put(_ ring: String, payload: String) throws -> RocheId {
        try put(ring, payload: Data(payload.utf8))
    }

    public func putVec(_ ring: String, payload: Data, vector: [Float]) throws -> RocheId {
        try withHandle { handle in
            try ring.withCString { cRing in
                var id = roche_id()
                let rc = payload.withUnsafeBytes { payloadBytes in
                    vector.withUnsafeBufferPointer { vectorBytes in
                        roche_put_vec(
                            handle,
                            cRing,
                            payloadBytes.baseAddress,
                            payload.count,
                            vectorBytes.baseAddress,
                            vector.count,
                            &id
                        )
                    }
                }
                try check(rc)
                return RocheId(c: id)
            }
        }
    }

    public func putVec(_ ring: String, payload: String, vector: [Float]) throws -> RocheId {
        try putVec(ring, payload: Data(payload.utf8), vector: vector)
    }

    public func get(_ id: RocheId) throws -> Data? {
        try withHandle { handle in
            var len = 0
            guard let ptr = roche_get(handle, id.cValue, &len) else {
                let err = Self.lastError()
                if case let .cAbi(message) = err, message.contains("not found") {
                    return nil
                }
                throw err
            }
            defer { roche_free(ptr) }
            return Data(bytes: ptr, count: len)
        }
    }

    public func getString(_ id: RocheId) throws -> String? {
        guard let data = try get(id) else {
            return nil
        }
        guard let value = String(data: data, encoding: .utf8) else {
            throw RocheDBError.invalidUtf8
        }
        return value
    }

    public func batchGet(_ ids: [RocheId]) throws -> [Data] {
        try withHandle { handle in
            var rawIds = ids.map(\.cValue)
            let count = rawIds.count
            let result = rawIds.withUnsafeMutableBufferPointer { buffer in
                roche_batch_get(handle, buffer.baseAddress, count)
            }
            guard let result else {
                throw Self.lastError()
            }
            defer { roche_batch_get_free(result) }
            guard let values = result.pointee.values else {
                return []
            }
            return (0..<Int(result.pointee.len)).map { index in
                let value = values.advanced(by: index).pointee
                guard let data = value.data, value.len > 0 else {
                    return Data()
                }
                return Data(bytes: data, count: value.len)
            }
        }
    }

    public func query(_ id: RocheId, selection: String) throws -> Data {
        try withHandle { handle in
            try selection.withCString { cSelection in
                var len = 0
                guard let ptr = roche_query(handle, id.cValue, cSelection, &len) else {
                    throw Self.lastError()
                }
                defer { roche_free(ptr) }
                return Data(bytes: ptr, count: len)
            }
        }
    }

    public func queryString(_ id: RocheId, selection: String) throws -> String {
        let data = try query(id, selection: selection)
        guard let value = String(data: data, encoding: .utf8) else {
            throw RocheDBError.invalidUtf8
        }
        return value
    }

    public func retrieve(
        vector: [Float],
        ring: String = "",
        budget: Int32 = 8,
        topRings: Int32 = 0,
        focus: Int32 = 0
    ) throws -> RetrieveResult {
        try withHandle { handle in
            try ring.withCString { cRing in
                let result = vector.withUnsafeBufferPointer { vectorBytes in
                    roche_retrieve(
                        handle,
                        vectorBytes.baseAddress,
                        vector.count,
                        cRing,
                        budget,
                        topRings,
                        focus
                    )
                }
                guard let result else {
                    throw Self.lastError()
                }
                defer { roche_retrieve_free(result) }
                let raw = result.pointee
                let hits: [RocheHit]
                if let rawHits = raw.hits {
                    hits = (0..<Int(raw.len)).map { index in
                        let hit = rawHits.advanced(by: index).pointee
                        let payload: Data
                        if let ptr = hit.payload, hit.payload_len > 0 {
                            payload = Data(bytes: ptr, count: hit.payload_len)
                        } else {
                            payload = Data()
                        }
                        return RocheHit(id: RocheId(c: hit.id), score: hit.score, payload: payload)
                    }
                } else {
                    hits = []
                }
                return RetrieveResult(
                    hits: hits,
                    stats: RetrieveStats(
                        totalVectors: Int(raw.total_vectors),
                        scanned: Int(raw.scanned),
                        skippedVectors: Int(raw.skipped_vectors),
                        returned: Int(raw.returned),
                        ringsTouched: Int(raw.rings_touched),
                        payloadBytes: Int(raw.payload_bytes),
                        estimatedTokens: Int(raw.estimated_tokens),
                        fanoutNodes: Int(raw.fanout_nodes),
                        candidateReduction: raw.candidate_reduction
                    )
                )
            }
        }
    }

    public func atlas(queryVector: [Float] = [], maxCentroidDims: Int32 = 8) throws -> String {
        try withHandle { handle in
            var len = 0
            let ptr = queryVector.withUnsafeBufferPointer { buffer in
                roche_atlas(handle, buffer.baseAddress, queryVector.count, maxCentroidDims, &len)
            }
            guard let ptr else {
                throw Self.lastError()
            }
            defer { roche_free(ptr) }
            guard let value = String(data: Data(bytes: ptr, count: len), encoding: .utf8) else {
                throw RocheDBError.invalidUtf8
            }
            return value
        }
    }

    public func locate(_ id: RocheId, at: Double = -1.0) throws -> Int {
        try withHandle { handle in
            let node = roche_locate(handle, id.cValue, at)
            if node < 0 {
                throw Self.lastError()
            }
            return Int(node)
        }
    }

    public func nextVisit(_ id: RocheId, node: Int32) throws -> Double {
        try withHandle { handle in
            let time = roche_next_visit(handle, id.cValue, node)
            if time < 0 {
                throw Self.lastError()
            }
            return time
        }
    }

    public func nextJoin(_ a: RocheId, _ b: RocheId) throws -> Double? {
        try withHandle { handle in
            let time = roche_next_join(handle, a.cValue, b.cValue)
            return time < 0 ? nil : time
        }
    }

    private static func ensureRuntime() throws {
        _ = runtimeInit
        let actual = roche_abi_version()
        if actual != abiVersion {
            throw RocheDBError.abiVersionMismatch(expected: abiVersion, actual: actual)
        }
    }

    private static func lastError() -> RocheDBError {
        guard let ptr = roche_last_error() else {
            return .cAbi("RocheDB C ABI error")
        }
        let message = String(cString: ptr)
        return .cAbi(message.isEmpty ? "RocheDB C ABI error" : message)
    }

    private func withHandle<T>(_ body: (UnsafeMutableRawPointer) throws -> T) throws -> T {
        guard let handle else {
            throw RocheDBError.closed
        }
        return try body(handle)
    }

    private func check(_ code: Int32) throws {
        if code != ROCHE_OK {
            throw Self.lastError()
        }
    }
}

private extension RocheId {
    init(c: roche_id) {
        self.init(parent: c.parent, epoch: c.epoch, seq: c.seq, tWrite: c.t_write)
    }

    var cValue: roche_id {
        roche_id(parent: parent, epoch: epoch, seq: seq, t_write: tWrite)
    }
}
